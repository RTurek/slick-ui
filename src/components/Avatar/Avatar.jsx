import React from 'react';
import MaterialAvatar from '@material-ui/core/Avatar';
import { withStyles } from '@material-ui/core/styles';
import PropTypes from 'prop-types';
import { COLORS } from '../../styles/colors';
import Anonymous from '../../assets/svg/Anonymous';
import { hasFullName, getFirstLetter } from '../../utils/string';
import {
  AVATAR_SIZE_OPTIONS,
  AVATAR_SIZES,
  AVATAR_VIEWBOX_SIZES,
  AVATAR_FONT_SIZES
} from '../../styles/sizes';

// TODO: Add a border with a few color variants
// TODO: refactor to be a functional component

function buildStyle(props) {
  const backgroundColor = COLORS.primary;
  const textColor = COLORS.white;
  const style = {
    color: textColor,
    height: `${AVATAR_SIZES.get(props.size)}`,
    width: `${AVATAR_SIZES.get(props.size)}`,
    margin: props.margin ? props.margin : 'auto',
    fontSize: `${AVATAR_FONT_SIZES.get(props.size)}`,
    backgroundColor
  };
  return style;
}

class Avatar extends React.PureComponent {
  constructor(props) {
    super(props);
    this.state = {
      error: false
    };
    this.styles = () => ({ root: buildStyle(this.props) });
  }

  handleError = (e) => {
    this.props.onError(e);
    this.setState({ error: true });
  };

  buildAvatar = () => {
    /* If there is an image URL, show the image.
     * If there is an error with the image, show the initials
     * If there are no initials (no first name OR no last name) show the default svg */
    const hasNoImageOrImageHasError = (!this.props.imageUrl || this.state.error) === true;
    const showImageAvatar = (this.props.imageUrl && !this.state.error) === true;
    const showInitialsAvatar =
      this.props.defaultToInitials &&
      hasNoImageOrImageHasError &&
      hasFullName(this.props.firstName, this.props.lastName);

    if (showImageAvatar) {
      const altInitials = '-';
      this.MaterialAvatar = withStyles(this.styles)((props) => {
        const { classes } = props;
        const height = AVATAR_SIZES.get(props.size);
        const width = height;
        return (
          <MaterialAvatar
            className="react-css"
            classes={{
              root: classes.root
            }}
            imgProps={{
              height,
              width
            }}
            onError={this.handleError}
            alt={altInitials}
            src={props.imageUrl}
            style={props.style}
          />
        );
      });
    } else if (showInitialsAvatar) {
      this.MaterialAvatar = withStyles(this.styles)((props) => {
        const { classes } = props;
        const initials = `${getFirstLetter(this.props.firstName)}${getFirstLetter(this.props.lastName)}`;
        return (
          <MaterialAvatar
            className="react-css"
            classes={{
              root: classes.root
            }}
            imgProps={{
              height: AVATAR_SIZES.get(props.size),
              width: AVATAR_SIZES.get(props.size)
            }}
            alt=""
            style={props.style}
          >
            {initials}
          </MaterialAvatar>
        );
      });
    } else {
      this.MaterialAvatar = withStyles(this.styles)(props => (
        <Anonymous
          fill={COLORS.primary}
          height={AVATAR_SIZES.get(props.size)}
          width={AVATAR_SIZES.get(props.size)}
          viewBox={AVATAR_VIEWBOX_SIZES.get(props.size)}
        />
      ));
    }
  };

  render() {
    this.buildAvatar();
    return <this.MaterialAvatar {...this.props} />;
  }
}

Avatar.propTypes = {
  /** url of avatar image */
  imageUrl: PropTypes.string,
  /** first name */
  firstName: PropTypes.string,
  /** last name */
  lastName: PropTypes.string,
  /** size options */
  size: PropTypes.oneOf(AVATAR_SIZE_OPTIONS),
  /** custom margin */
  margin: PropTypes.string,
  /** show initials as default if image doesn't exist  */
  defaultToInitials: PropTypes.bool,
  /** onError function */
  onError: PropTypes.func
};

Avatar.defaultProps = {
  firstName: null,
  lastName: null,
  imageUrl: null,
  size: 'small',
  margin: 'auto',
  defaultToInitials: true,
  onError: () => {}
};

export default Avatar;
