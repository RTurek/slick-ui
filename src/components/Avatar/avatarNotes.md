#### Avatar
---
This variant on Material UI's Avatar gives you some control over color, size, and lets you add a fallback icon.

#### Notes
---
- If there is an image URL, we show the image.
- If there is an error with the image, we show the initials.
- If there are no initials (no first name OR no last name) we show the default svg (the Anonymous component).

#### Sources
---
- https://material-ui.com/demos/avatars/
