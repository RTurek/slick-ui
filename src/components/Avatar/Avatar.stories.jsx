import React from 'react';
import { withInfo } from '@storybook/addon-info';
import { select, text } from '@storybook/addon-knobs';
import AccountCircle from '@material-ui/icons/AccountCircle';
import Avatar from './Avatar';
import StoryWrapper from '../../styles/StoryWrapper';
import { storybookStyles } from '../../../.storybook/storybookTheming';
import avatarNotes from './avatarNotes.md';

const defaultSize = 'small';

const sizeOptions = {
  xxsmall: 'xxsmall',
  xsmall: 'xsmall',
  small: 'small',
  medium: 'medium',
  large: 'large'
};

export default {
  title: 'Core / Avatar',

  decorators: [
    withInfo({
      text: avatarNotes,
      header: true,
      inline: true,
      styles: storybookStyles,
      source: false,
      propTables: [Avatar],
      propTablesExclude: [Avatar, AccountCircle],
      excludedPropTypes: ['theme']
    })
  ]
};

export const sandbox = () => (
  <StoryWrapper>
    <Avatar
      firstName={text('First Name', 'Sandbox')}
      lastName={text('Last Name', 'User')}
      imageUrl={text('Avatar Source URL', 'http://placekitten.com/g/60/60')}
      size={select('Size', sizeOptions, defaultSize)}
    />
  </StoryWrapper>
);

sandbox.story = {
  name: 'Sandbox',
  parameters: { avatarNotes }
};

export const withImage = () => (
  <StoryWrapper>
    <Avatar
      firstName="Erwin"
      lastName="Schrödinger"
      imageUrl="http://placekitten.com/60/60"
      size={select('Size', sizeOptions, defaultSize)}
    />
  </StoryWrapper>
);

withImage.story = {
  name: 'With Image',
  parameters: { avatarNotes }
};

export const withInitials = () => (
  <StoryWrapper>
    <Avatar
      firstName="Erwin"
      lastName="Schrödinger"
      size={select('Size', sizeOptions, defaultSize)}
    />
  </StoryWrapper>
);

withInitials.story = {
  name: 'With Initials',
  parameters: { avatarNotes }
};

// export const withSeasonalBadge = () => (
//   <StoryWrapper>
//     <Badge invisible={false} customBadgeVariant="seasonal-badge-left">
//       <Avatar
//         firstName={text('First Name', 'Santa')}
//         lastName={text('Last Name', 'Claus')}
//         imageUrl={text('Avatar Source URL', 'https://media.bizj.us/view/img/1586581/bloombergphotowillferrellelf*1200xx1388-777-0-354.jpg')}
//         size={select('Size', sizeOptions, 'xxsmall')}
//       />
//     </Badge>
//   </StoryWrapper>
// );

// withSeasonalBadge.story = {
//   name: 'With Seasonal Badge',
//   parameters: { avatarNotes }
// };

export const placeholderSvg = () => (
  <StoryWrapper>
    <Avatar size={select('Size', sizeOptions, defaultSize)} />
  </StoryWrapper>
);

placeholderSvg.story = {
  name: 'Placeholder SVG',
  parameters: { avatarNotes }
};
