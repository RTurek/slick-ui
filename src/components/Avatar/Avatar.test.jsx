import React from 'react';
import renderer from 'react-test-renderer';
import Avatar from './Avatar';

const fullProps = {
  imageUrl: 'http://i.pravatar.cc/300',
  firstName: 'John',
  lastName: 'Max'
};
const hasInitialsProps = {
  firstName: 'John',
  lastName: 'Max'
};
const anonProps = {};

test('Avatar should render', () => {
  const component = (
    <Avatar {...fullProps} />
  );
  const tree = renderer.create(component).toJSON();
  expect(tree).toMatchSnapshot();
});

test('Avatar should render Initials when there is no image', () => {
  const component = (
    <Avatar {...hasInitialsProps} />
  );
  const tree = renderer.create(component).toJSON();
  expect(tree).toMatchSnapshot();
});

test('Avatar should render Anonymous when there is no image or full name', () => {
  const component = (
    <Avatar {...anonProps} />
  );
  const tree = renderer.create(component).toJSON();
  expect(tree).toMatchSnapshot();
});
