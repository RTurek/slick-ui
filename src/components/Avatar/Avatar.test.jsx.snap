// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avatar should render 1`] = `
<div
  className="MuiAvatar-root Component-root-1 MuiAvatar-circle react-css"
  onError={[Function]}
>
  <img
    alt="-"
    className="MuiAvatar-img"
    height="60px"
    src="http://i.pravatar.cc/300"
    width="60px"
  />
</div>
`;

exports[`Avatar should render Anonymous when there is no image or full name 1`] = `
<svg
  className="svg-icon "
  height="60px"
  onClick={[Function]}
  style={Object {}}
  viewBox="0 0 64 64"
  width="60px"
  xmlns="http://www.w3.org/2000/svg"
  xmlnsXlink="http://www.w3.org/1999/xlink"
>
  <path
    d="M0 32C0 14.327 14.327 0 32 0c17.673 0 32 14.327 32 32 0 17.673-14.327 32-32 32C14.327 64 0 49.673 0 32zm9.35 16.24C14.407 55.285 22.666 59.872 32 59.872c9.333 0 17.592-4.587 22.65-11.63a8.516 8.516 0 0 0-6.391-7.273l-4.178-1.044A18.441 18.441 0 0 1 32 44.387c-4.48 0-8.719-1.575-12.081-4.463l-4.178 1.044a8.516 8.516 0 0 0-6.392 7.273v-.001zm8.2-22.433c0 7.981 6.469 14.452 14.45 14.452 7.981 0 14.452-6.471 14.452-14.454 0-7.98-6.47-14.45-14.452-14.45-7.981 0-14.452 6.47-14.452 14.452h.001z"
    fill="#002856"
  />
</svg>
`;

exports[`Avatar should render Initials when there is no image 1`] = `
<div
  className="MuiAvatar-root Component-root-2 MuiAvatar-circle react-css MuiAvatar-colorDefault"
>
  JM
</div>
`;
